File: 104/Main.java
import java.io.*;
import java.util.*;

public class Main {
   public static void main(String argv[]) throws IOException {
      String line;
      StringTokenizer tokenizer;
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      while ((line = in.readLine()) != null) {
         int n = Integer.parseInt(line);
         double[][][] profit = new double[n][n][n];
         for (int i = 0; i < n; i++) {
            line = in.readLine();
            tokenizer = new StringTokenizer(line);
            for (int j = 0; j < n; j++) {
               if (i == j) {
                  profit[0][i][j] = 1.0;
               } else {
                  profit[0][i][j] = Double.parseDouble(tokenizer.nextToken());
               }
            }
         }
         findArbitrageSequence(profit);
      }
   }

   public static void findArbitrageSequence(double[][][] profit) {
      int n = profit.length;
      int[][][] path = new int[n][n][n];

      for (int i = 0; i < n; i++) {
         for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
               path[k][i][j] = j;
            }
         }
      }

      for (int m = 1; m < n; m++) {
         for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
               for (int j = 0; j < n; j++) {
                  double best = profit[m-1][i][k] * profit[0][k][j];
                  if (best > profit[m][i][j]) {
                     profit[m][i][j] = best;
                     path[m][i][j] = k;
                  }
               }
            }
         }
      }

      int steps = 0;
      for (int i = 1; i < n; i++) {
         for (int j = 0; j < n; j++) {
            if (profit[i][j][j] >= 1.01) {
               n = j;
               steps = i;
            }
         }
      }
      if (steps > 0) {
         int start = n, end = n;
         int[] step = new int[steps + 1];
         for (int s = steps; s >= 0; s--) {
            step[s] = start + 1;
            start = path[s][end][start];
         }
         for (int i = 0; i < step.length; i++)
            System.out.print(step[i] + " ");
         System.out.println(start + 1);
      } else {
         System.out.println("no arbitrage sequence exists");
      }
   }
}

File: 104/main.c
#include <stdio.h>
#include <memory.h>

int find_arbitrage_sequence(int n)
{
   int i, j, k, m;
   int step[20], path[20][20][20];
   double profit, best[20][20][20];

   memset(best, 0, sizeof(best));

   for (i = 0; i < n; i++) {
      for (j = 0; j < n; j++) {
         if (i == j) {
            best[0][i][j] = 1.0;
         } else {
            scanf("%lf", &best[0][i][j]);
         }
      }
   }

   for (k = 0; k < n; k++) {
      for (i = 0; i < n; i++) {
         for (j = 0; j < n; j++) {
            path[k][i][j] = j;
         }
      }
   }

   for (m = 1; m < n; m++) {
      for (k = 0; k < n; k++) {
         for (i = 0; i < n; i++) {
            for (j = 0; j < n; j++) {
               profit = best[m-1][i][k] * best[0][k][j];
               if (best[m][i][j] < profit) {
                  best[m][i][j] = profit;
                  path[m][i][j] = k;
               }
            }
         }
      }
   }

   k = 0;
   for (m = 1; m < n && k == 0; m++) {
      for (i = 0; i < n && k == 0; i++) {
         if (best[m][i][i] >= 1.01) {
            k = m;
         }
      }
   }

   if (k > 0) {
      j = i;
      for (m = k; m >= 0; m--) {
         step[m] = i;
         i = path[m][j][i];
      }
      for (m = 0; m <= k; m++) {
         printf("%d ", step[m] + 1);
      }
      printf("%d\n", i + 1);
   } else {
      printf("no arbitrage sequence exists\n");
   }

   return 0;
}

int main(int argc, char* argv[])
{
   int n;

   while (scanf("%d", &n) == 1) {
      find_arbitrage_sequence(n);
   }

   return 0;
}
